replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,259 @@
+   * if necessary.
+   */
+  @Override
+  public void run() {
+    while (running || !Thread.currentThread().isInterrupted()) {
+      try {
+        Task task = queue.take();
+        this.processing.set(true);
+        if (supported(task)) {
+          this.handle(task);
+        }
+        this.processing.set(false);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Set the running flag to false and wait for the thread to die. If an item is currently in the
+   * process of transferring, it will finish.
+   */
+  public void cancel() {
+    this.running = false;
+    try {
+      Thread.currentThread().join();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  /**
+   * @return {@code true} if this worker is transferring an item, otherwise {@code false}
+   */
+  public boolean isProcessing() {
+    return processing.get();
+  }
+
+  private boolean supported(Task task) throws InterruptedException {
+    if (!OperationType.UNKNOWN.equals(task.getOperation())) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          String.format("Operation task %s is not supported by this worker", task));
+      return false;
+    }
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    boolean proceed = checkAdapterAvailabilities(task);
+    if (!proceed) {
+      return;
+    }
+
+    List<DDFMetadataInfo> infos = task.metadatas()
+        .filter(t -> t instanceof DDFMetadataInfo)
+        .map(DDFMetadataInfo.class::cast)
+        .collect(Collectors.toList());
+
+    if (infos.size() > 1) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          "more than one metadata is currently not supported by this worker");
+      return;
+    }
+
+    if(infos.isEmpty()) {
+      task.fail(ErrorCode.UNSUPPORTED_ERROR,
+          "unable to handle any of the metadatas received in this task");
+      return;
+    }
+
+    final DDFMetadataInfo info = infos.get(0);
+    final Metadata metadata = new MetadataImpl(((Map) info.getData()).get("raw-metadata"),
+        info.getDataClass(), task.getId(),
+        Date.from(info.getLastModified()));
+
+    final OperationType type = task.getOperation();
+    switch (type) {
+      case HARVEST:
+      case REPLICATE_FROM:
+        doCreate(task, metadata, destination, local);
+        break;
+      case REPLICATE_TO:
+        doCreate(task, metadata, local, destination);
+        break;
+      case DELETE:
+        // always delete on non-local site
+        doDelete(task, metadata);
+        break;
+      default:
+        // should never hit this since deserialization will return UNKNOWN
+        task.fail(ErrorCode.UNSUPPORTED_ERROR,
+            String.format("Operation task %s is not supported by this worker", task));
+    }
+  }
+
+  private void doCreate(final Task task, final Metadata metadata, final NodeAdapter source,
```

Comment:
❗️ Exception handling when calling `exist()`, `createRequest`, and all other methods on the node adapter needs to happen somewhere. We also need to handle properly interruption caused by the manager trying to stop the worker which could translate into `InterruptedIOException` which needs to be properly reported by the adapters. And that I don't think can be pushed off to a different story.